yarn/src/main/scala/org/apache/spark/deploy/yarn/AMDelegationTokenRenewer.scala

```diff
@@ -0,0 +1,201 @@
+      val keytab = sparkConf.get("spark.yarn.keytab")
+
+      def getRenewalInterval = {
+        val credentials = UserGroupInformation.getCurrentUser.getCredentials
+        math.max((0.75 * (hadoopUtil.getLatestTokenValidity(credentials) -
+          System.currentTimeMillis())).toLong, 0L)
+      }
+
+      def scheduleRenewal(runnable: Runnable) = {
+        val renewalInterval = getRenewalInterval
+        logInfo(s"Scheduling login from keytab in $renewalInterval millis.")
+        delegationTokenRenewer.schedule(runnable, renewalInterval, TimeUnit.MILLISECONDS)
+      }
+
+      // This thread periodically runs on the driver to update the delegation tokens on HDFS.
+      val driverTokenRenewerRunnable =
+        new Runnable {
+          override def run(): Unit = {
+            try {
+              writeNewTokensToHDFS(principal, keytab)
+            } catch {
+              case e: Exception =>
+                logWarning("Failed to write out new credentials to HDFS, will try again in an " +
+                  "hour! If this happens too often tasks will fail.", e)
+                delegationTokenRenewer.schedule(this, 1, TimeUnit.HOURS)
+                return
+            }
+            cleanupOldFiles()
+            scheduleRenewal(this)
+          }
+        }
+      // If this is an AM restart, it is possible that the original tokens have expired, which
+      // means we need to login immediately to get new tokens.
+      if (getRenewalInterval == 0) writeNewTokensToHDFS(principal, keytab)
+      // Schedule update of credentials
+      scheduleRenewal(driverTokenRenewerRunnable)
+
+    }
+  }
+
+  // Keeps only files that are newer than 30 days, and deletes everything else. At least 5 files
+  // are kept for safety
+  private def cleanupOldFiles(): Unit = {
+    import scala.concurrent.duration._
+    try {
+      val remoteFs = FileSystem.get(hadoopConf)
+      val credentialsPath = new Path(sparkConf.get("spark.yarn.credentials.file"))
+      hadoopUtil.listFilesSorted(
+        remoteFs, credentialsPath.getParent, credentialsPath.getName, ".tmp").dropRight(5)
+        .takeWhile(_.getModificationTime < System.currentTimeMillis() - (30 days).toMillis)
+        .foreach(x => remoteFs.delete(x.getPath, true))
+    } catch {
+      // Such errors are not fatal, so don't throw. Make sure they are logged though
+      case e: Exception =>
+        logWarning("Error while attempting to cleanup old tokens. If you are seeing many such " +
+          "warnings there may be an issue with your HDFS cluster.")
+    }
+  }
+
+  private def writeNewTokensToHDFS(principal: String, keytab: String): Unit = {
+    if (!loggedInViaKeytab) {
+      // Keytab is copied by YARN to the working directory of the AM, so full path is
+      // not needed.
+      logInfo(s"Attempting to login to KDC using principal: $principal")
+      loggedInUGI = UserGroupInformation.loginUserFromKeytabAndReturnUGI(
```

Comment:
I dont' think you need to use the AndReturnUGI part of this.  If you need the UGI you can always just use getCurrentUser.